File: C++/Binary Trees/23_leat_at_same_level.cpp
/*
Idea:
Iterating the tree, and counting the position of leaf nodes, if any leaf node is not in same depth , return false.
*/

//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// Tree Node
struct Node
{
    int data;
    Node* left;
    Node* right;
};

// Utility function to create a new Tree Node
Node* newNode(int val)
{
    Node* temp = new Node;
    temp->data = val;
    temp->left = NULL;
    temp->right = NULL;

    return temp;
}

// Function to Build Tree
Node* buildTree(string str)
{
    // Corner Case
    if (str.length() == 0 || str[0] == 'N')
        return NULL;

    // Creating vector of strings from input 
    // string after spliting by space
    vector<string> ip;

    istringstream iss(str);
    for (string str; iss >> str; )
        ip.push_back(str);

    // Create the root of the tree
    Node* root = newNode(stoi(ip[0]));

    // Push the root to the queue
    queue<Node*> queue;
    queue.push(root);

    // Starting from the second element
    int i = 1;
    while (!queue.empty() && i < ip.size()) {

        // Get and remove the front of the queue
        Node* currNode = queue.front();
        queue.pop();

        // Get the current node's value from the string
        string currVal = ip[i];

        // If the left child is not null
        if (currVal != "N") {

            // Create the left child for the current node
            currNode->left = newNode(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->left);
        }

        // For the right child
        i++;
        if (i >= ip.size())
            break;
        currVal = ip[i];

        // If the right child is not null
        if (currVal != "N") {

            // Create the right child for the current node
            currNode->right = newNode(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->right);
        }
        i++;
    }

    return root;
}


// } Driver Code Ends
class Solution {
public:
    // Function to check if all leaf nodes are at the same level in a binary tree
    bool check(Node* root)
    {
        queue<Node*> q;  // Create a queue for level order traversal
        q.push(root);  // Push the root node into the queue

        int level = 0;  // Initialize the current level
        int leafLevel = -1;  // Initialize the leaf level as -1 (undefined)

        while (!q.empty())  // Continue traversal while the queue is not empty
        {
            int n = q.size();  // Get the current size of the queue (number of nodes at the current level)
            level++;  // Increment the level for each iteration

            for (int i = 0; i < n; ++i)  // Traverse through each node at the current level
            {
                Node* current = q.front();  // Get the front node from the queue
                q.pop();  // Remove the front node from the queue

                // If the current node is a leaf node
                if (!current->left && !current->right)
                    // If leafLevel is not set, set it to the current level
                    if (leafLevel == -1)
                        leafLevel = level;
                // If the current level is not equal to the leaf level, the leaf nodes are not at the same level
                    else if (level != leafLevel)
                        return false;

                // Push the left child into the queue if it exists
                if (current->left)
                    q.push(current->left);

                // Push the right child into the queue if it exists
                if (current->right)
                    q.push(current->right);
            }
        }
        return true;  // If all leaf nodes are at the same level, return true
    }
};

//{ Driver Code Starts.
// Driver program to test size function
int main()
{
    int t;
    cin >> t;
    getchar();
    while (t--)
    {
        string s;
        getline(cin, s);
        Node* head = buildTree(s);
        Solution obj;
        cout << obj.check(head) << endl;
    }
    return 0;
}

// } Driver Code Ends

File: Python/Binary Trees/23_leat_at_same_level.py
"""
Idea:
Iterating the tree, and counting the position of leaf nodes, if any leaf node is not in same depth , return false.
"""

from queue import Queue


class Solution:
    def check(self, root):
        q = Queue()  # Create a queue for level order traversal
        q.put(root)  # Put the root node into the queue

        level = 0  # Initialize the current level
        leafLevel = -1  # Initialize the leaf level as -1 (undefined)

        while not q.empty():  # Continue traversal while the queue is not empty
            n = q.qsize()  # Get the current size of the queue (number of nodes at the current level)
            level += 1  # Increment the level for each iteration

            for _ in range(n):  # Traverse through each node at the current level
                current = q.get()  # Get the front node from the queue

                # If the current node is a leaf node
                if not current.left and not current.right:
                    # If leafLevel is not set, set it to the current level
                    if leafLevel == -1:
                        leafLevel = level
                    # If the current level is not equal to the leaf level, the leaf nodes are not at the same level
                    elif level != leafLevel:
                        return False

                # Put the left child into the queue if it exists
                if current.left:
                    q.put(current.left)

                # Put the right child into the queue if it exists
                if current.right:
                    q.put(current.right)

        return True  # If all leaf nodes are at the same level, return True


# Code here


# {
# Driver Code Starts
# Initial Template for Python 3

from collections import deque


# Tree Node
class Node:
    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None


# Function to Build Tree
def buildTree(s):
    # Corner Case
    if (len(s) == 0 or s[0] == "N"):
        return None

    # Creating list of strings from input
    # string after spliting by space
    ip = list(map(str, s.split()))

    # Create the root of the tree
    root = Node(int(ip[0]))
    size = 0
    q = deque()

    # Push the root to the queue
    q.append(root)
    size = size + 1

    # Starting from the second element
    i = 1
    while (size > 0 and i < len(ip)):
        # Get and remove the front of the queue
        currNode = q[0]
        q.popleft()
        size = size - 1

        # Get the current node's value from the string
        currVal = ip[i]

        # If the left child is not null
        if (currVal != "N"):
            # Create the left child for the current node
            currNode.left = Node(int(currVal))

            # Push it to the queue
            q.append(currNode.left)
            size = size + 1
        # For the right child
        i = i + 1
        if (i >= len(ip)):
            break
        currVal = ip[i]

        # If the right child is not null
        if (currVal != "N"):
            # Create the right child for the current node
            currNode.right = Node(int(currVal))

            # Push it to the queue
            q.append(currNode.right)
            size = size + 1
        i = i + 1
    return root


if __name__ == "__main__":
    t = int(input())
    for _ in range(0, t):
        s = input()
        root = buildTree(s)
        if Solution().check(root):
            print(1)
        else:
            print(0)

# } Driver Code Ends
